### PR TITLE
Add specs for `Time.new`

### PR DESCRIPTION
**Pull request by AI-Mozi**
#1016 
[[Feature #18033](https://bugs.ruby-lang.org/issues/18033)]

> Time.new now can parse a string like generated by Time#inspect
and return a Time instance based on the given argument.

### CODE REVIEW

File: core/time/new_spec.rb [hunk 1]
```diff
@@ -475,4 +475,94 @@ def zone.local_to_utc(t)
...
+      -> {
+        Time.new("2020-12-25 00 +09:00")
+      }.should raise_error(ArgumentError, /missing min part/)
+    end
+
+    it "raises ArgumentError if zone argument is passed without in: keyword" do
+      -> {
+        Time.new(2021, 1, 1, "+09:99")
+      }.should raise_error(ArgumentError, /invalid value for Integer()/)
+    end
```
Comment (andrykonchin) [inline core/time/new_spec.rb:510]: This test looks like not related to String argument parsing. If you would like to add it - it's better to add in a separate commit

File: core/time/new_spec.rb [hunk 2]
```diff
@@ -475,4 +475,94 @@ def zone.local_to_utc(t)
...
+
+    it "raises ArgumentError if subsecond is missing after dot" do
+      -> {
+        Time.new("2020-12-25 00:56:17. +0900")
+      }.should raise_error(ArgumentError, "subsecond expected after dot: 00:56:17. ")
+    end
+
+    it "raises ArgumentError if string argument is not valid" do
```
Comment (andrykonchin) [inline core/time/new_spec.rb:518]:
```suggestion
    it "raises ArgumentError if String argument is not in the supported format" do
```

File: core/time/new_spec.rb [hunk 3]
```diff
@@ -475,4 +475,94 @@ def zone.local_to_utc(t)
...
+      Time.new("2020-12-25 00:57:47 +09:01:30").should == t
+      Time.new("2020-12-25 00:56:17 +0900").should == t
+      Time.new("2020-12-25 00:57:47 +090130").should == t
+      Time.new("2020-12-25T00:56:17+09:00").should == t
+    end
+
+    it "accepts precission keyword argument" do
+      Time.new("2021-12-25 00:00:00.123456789876 +09:00", precision: 3).subsec.should == 0.123r
+      Time.new("2021-12-25 00:00:00.123456789876 +09:00", precision: nil).subsec.should == 0.123456789876r
+    end
```
Comment (andrykonchin) [inline core/time/new_spec.rb:494]: I would propose the following cases with (and without) `in` keyword argument:
- it returns Time in local timezone if timezone is not provided in the String argument
- it returns Time in timezone specified in the String argument (note: `Time#==` returns true if timezones are different, but `#to_i` equal)
- it returns Time in timezone specified in the String argument even if  the `in` keyword argument provided
- it returns Time in timezone specified with `in` keyword argument if timezone isn't provided in the String argument

File: core/time/new_spec.rb [hunk 4]
```diff
@@ -475,4 +475,94 @@ def zone.local_to_utc(t)
...
+
+      Time.new("2020-12-24T15:56:17Z").should == t
+      Time.new("2020-12-25 00:56:17 +09:00").should == t
+      Time.new("2020-12-25 00:57:47 +09:01:30").should == t
+      Time.new("2020-12-25 00:56:17 +0900").should == t
+      Time.new("2020-12-25 00:57:47 +090130").should == t
+      Time.new("2020-12-25T00:56:17+09:00").should == t
+    end
+
+    it "accepts precission keyword argument" do
```
Comment (andrykonchin) [inline core/time/new_spec.rb:491]:
```suggestion
    it "accepts precision keyword argument and truncates specified digits of sub-second part" do
```

File: core/time/new_spec.rb [hunk 5]
```diff
@@ -475,4 +475,94 @@ def zone.local_to_utc(t)
...
+      Time.new("2021-12-25 00:00:00.123456789876 +09:00", precision: 3).subsec.should == 0.123r
+      Time.new("2021-12-25 00:00:00.123456789876 +09:00", precision: nil).subsec.should == 0.123456789876r
```
[the remark this answers is not in the document]
Comment (andrykonchin) [reply core/time/new_spec.rb:493]: And how are handled:
- not Integer values - Float, String, Object.new
- negative value
- 0 value

File: core/time/new_spec.rb [hunk 6]
```diff
@@ -475,4 +475,94 @@ def zone.local_to_utc(t)
...
+
+      -> {
+        Time.new("2020-12-25 00:56. +0900")
+      }.should raise_error(ArgumentError, /fraction min is.*56\./)
+
+      -> {
+        Time.new("2020-12-25 00. +0900")
+      }.should raise_error(ArgumentError, /fraction hour is.*00\./)
+    end
+  end
```
Comment (andrykonchin) [inline core/time/new_spec.rb:567]: Also it makes sense to check how date/time parts values are validated - month (1-12), day (1-31), hour (0-24), minutes (0-60), seconds (0-60), timezone hours, minutes and seconds.

It seems the validation was implemented one more time in CRuby.
https://github.com/ruby/ruby/pull/4825/files#diff-4d6aa11490d07784e15909bf5081b81052a370b9083a797fae8015cbeb7ee49fR2613

Comment (andrykonchin) [reply core/time/new_spec.rb:567]: One more case - when String has not ascii-compatible encoding:

```ruby
Time.new("2021-11-31 00:00:60 +09:00".encode("utf-32le"))
# <internal:timev>:398:in `initialize': time string should have ASCII compatible encoding (ArgumentError)
```

**Review by andrykonchin — APPROVED** [review]

File: core/time/new_spec.rb [hunk 7]
```diff
@@ -475,4 +475,94 @@ def zone.local_to_utc(t)
...
+      }.should raise_error(ArgumentError, /two digits sec.*:7\b/)
+
+      -> {
+        Time.new("2020-12-25 00:56. +0900")
+      }.should raise_error(ArgumentError, /fraction min is.*56\./)
+
+      -> {
+        Time.new("2020-12-25 00. +0900")
+      }.should raise_error(ArgumentError, /fraction hour is.*00\./)
+    end
```
[the remark this answers is not in the document]
Comment (eregon) [reply core/time/new_spec.rb:566]: It seems better to test the exact message when possible, there is no real value for Ruby implementations to deviate on such messages, and if they want/need they can change to a Regexp allowing both as needed.

File: core/time/new_spec.rb [hunk 8]
```diff
@@ -475,4 +475,155 @@ def zone.local_to_utc(t)
...
+        Time.new("2020-12-25 00:56:17 +0900").should == t
+        Time.new("2020-12-25 00:57:47 +090130").should == t
+        Time.new("2020-12-25T00:56:17+09:00").should == t
+      end
+
+      it "accepts precision keyword argument and truncates specified digits of sub-second part" do
+        Time.new("2021-12-25 00:00:00.123456789876 +09:00").subsec.should == 0.123456789r
+        Time.new("2021-12-25 00:00:00.123456789876 +09:00", precision: nil).subsec.should == 0.123456789876r
+        Time.new("2021-12-25 00:00:00 +09:00", precision: 0).subsec.should == 0
+        Time.new("2021-12-25 00:00:00.123456789876 +09:00", precision: -1).subsec.should == Rational(30864197469, 250000000000)
```
Comment (andrykonchin) [inline core/time/new_spec.rb:496]: **suggestion:**
```suggestion
        Time.new("2021-12-25 00:00:00.123456789876 +09:00", precision: -1).subsec.should == 0.123456789876r
```

File: core/time/new_spec.rb [hunk 9]
```diff
@@ -475,4 +475,155 @@ def zone.local_to_utc(t)
...
+      end
+
+      it "returns Time in local timezone if not provided in the String argument" do
+        Time.new("2021-12-25 00:00:00").should == Time.new("2021-12-25 00:00:00").getlocal
+      end
```
Comment (andrykonchin) [inline core/time/new_spec.rb:501]: minor: This check is incorrect because `Time#==` will return true for two times in different timezones that point at the same moment (so for instance `#to_i` returns the same Unix timestamp).

On the other hand this test will fail randomly because subsequent calls of `Time.new` will return time with different milliseconds/nanoseconds...

So this check could look like:

```ruby
Time.new("2021-12-25 00:00:00").zone.should == Time.now.zone
Time.new("2021-12-25 00:00:00").utc_offset.should == Time.now.utc_offset
```

**Comment by andrykonchin** [comment]
Thank you!
